### PR TITLE
Rename WeakForm to Functional

### LINKED ISSUE #464: Rename/move `variational/weak_form` to `functional`
Does the directory "variational_form" still make sense? I vaguely remember a conversation that this was renamed to weak form ...

_Originally posted by @white238 in https://github.com/LLNL/serac/pull/418#discussion_r633852855_

### PR DESCRIPTION
**Pull request by jamiebramwell**
This renames the `WeakForm` class to `Functional`.

Resolves https://github.com/LLNL/serac/issues/464.

### CODE REVIEW

File: src/docs/doxygen/mainpage.md [hunk 1]
```diff
@@ -21,6 +21,7 @@
 * [First order ODE solver](@ref serac::mfem_ext::FirstOrderODE)
 * [Second order ODE solver](@ref serac::mfem_ext::SecondOrderODE)
 
-#### Weak Form Interface ####
+#### Functional Interface ####
 
 * [Integral](@ref serac::Integral)
+* <a style="font-weight:bold" href="classserac_1_1Functional_3_01test_07trial_08_4.html">Functional</a>
```
Comment (joshessman-llnl) [inline src/docs/doxygen/mainpage.md:27]: Nice!

**Review by joshessman-llnl — APPROVED** [review]
Ctrl + F "weak" didn't come up with anything suspicious, so looks good to me! 😄

**Review by samuelpmishLLNL — APPROVED** [review]

File: src/docs/doxygen/mainpage.md (hunk 1 above)
Comment (jamiebramwell) [reply src/docs/doxygen/mainpage.md:27]: Yeah, I don't like it and we need to make sure this link doesn't break, but hay, it works!

**Comment by jamiebramwell** [comment]
> 
> 
> Ctrl + F "weak" didn't come up with anything suspicious, so looks good to me! 😄

That was my strategy for this PR! 😀

**Pull request merged**